File: solutions/Leetcode_86/leetcode_86.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

using namespace std;

class Solution1
{
public:
    ListNode *partition(ListNode *head, int x)
    {
        ListNode head1(0), head2(0);
        ListNode *head1_ptr = &head1, *head2_ptr = &head2;

        while (head)
        {
            if (head->val < x)
            {
                head1_ptr->next = head;
                head1_ptr = head1_ptr->next;
            }
            else
            {
                head2_ptr->next = head;
                head2_ptr = head2_ptr->next;
            }
            head = head->next;
        }
        head2_ptr->next = nullptr;
        head1_ptr->next = head2.next;
        return head1.next;
    }
};

File: solutions/Leetcode_86/leetcode_86.py
from solutions.common import ListNode


class Solution1:
    def partition(self, head: ListNode, x: int) -> ListNode:
        head_1 = head_1_cp = ListNode(0)
        head_2 = head_2_cp = ListNode(0)
        while head:
            if head.val < x:
                head_1_cp.next = head
                head_1_cp = head_1_cp.next
            else:
                head_2_cp.next = head
                head_2_cp = head_2_cp.next
            head = head.next
        head_2_cp.next = None
        head_1_cp.next = head_2.next
        return head_1.next


class Solution2:
    def partition(self, head: ListNode, x: int) -> ListNode:
        dummy = ListNode(0, head)
        prev = dummy
        while head and head.val < x:
            prev, head = head, head.next
        temp = prev
        while head:
            if head.val < x:
                prev.next = head.next
                head.next = temp.next
                temp.next = head
                temp = temp.next
                head = prev.next
            else:
                prev = head
                head = head.next
        return dummy.next


if __name__ == '__main__':
    head = ListNode.from_list([1, 4, 3, 2, 5, 2])
    ans = Solution1().partition(head, 3)
    while ans:
        print(ans.val)
        ans = ans.next
